src/network.rs

```diff
@@ -0,0 +1,508 @@
+					_ = sender.send(Ok(())).ok();
+					return;
+				}
+
+				if let hash_map::Entry::Vacant(e) = self.pending_dials.entry(peer_id) {
+					match self
+						.swarm
+						.dial(peer_addr.with(Protocol::P2p(peer_id.into())))
+					{
+						Ok(()) => {
+							e.insert(sender);
+						},
+						Err(e) => {
+							_ = sender.send(Err(e.into()));
+						},
+					}
+				} else {
+					todo!("Implement logic for peer thats already beeing dialed.");
+				}
+			},
+			Command::Bootstrap { sender } => {
+				let query_id = self
+					.swarm
+					.behaviour_mut()
+					.kademlia
+					.bootstrap()
+					.expect("DHT not to be empty");
+
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::Bootstrap(sender));
+			},
+			Command::GetKadRecord {
+				key,
+				quorum,
+				sender,
+			} => {
+				let query_id = self.swarm.behaviour_mut().kademlia.get_record(key, quorum);
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::GetRecord(sender));
+			},
+			Command::PutKadRecord {
+				record,
+				quorum,
+				sender,
+			} => {
+				let query_id = self
+					.swarm
+					.behaviour_mut()
+					.kademlia
+					.put_record(record, quorum)
+					.expect("No put error.");
+
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::PutRecord(sender));
+			},
+		}
+	}
+}
+
+pub fn init(
+	seed: Option<u8>,
+	psk_path: &String,
+) -> Result<(Client, impl Stream<Item = Event>, EventLoop)> {
+	// Create a public/private key pair, either based on a seed or random
+	let id_keys = match seed {
+		Some(seed) => {
+			let mut bytes = [0u8; 32];
+			bytes[0] = seed;
+			let secret_key = ed25519::SecretKey::from_bytes(&mut bytes)
+				.expect("Error should only appear if length is wrong.");
+			identity::Keypair::Ed25519(secret_key.into())
+		},
+		None => identity::Keypair::generate_ed25519(),
+	};
+	let local_peer_id = id_keys.public().to_peer_id();
+	info!("Local peer id: {:?}", local_peer_id);
+
+	// try to get psk
+	let psk: Option<PreSharedKey> = get_psk(psk_path)?
+		.map(|text| PreSharedKey::from_str(&text))
+		.transpose()?;
+	// create transport
+	let transport = setup_transport(id_keys, psk);
+
+	// create swarm that manages peers and events
+	let swarm = {
+		let mut kad_cfg = KademliaConfig::default();
+		kad_cfg.set_query_timeout(Duration::from_secs(5 * 60));
+		let store_cfg = MemoryStoreConfig {
+			max_records: 24000000, // ~2hrs
+			max_value_bytes: 100,
+			max_providers_per_key: 1,
+			max_provided_keys: 100000,
+		};
+		let kad_store = MemoryStore::with_config(local_peer_id, store_cfg);
+		let behaviour = NetworkBehaviour {
+			kademlia: Kademlia::with_config(local_peer_id, kad_store, kad_cfg),
+		};
+
+		// Build the Swarm, connecting the lower transport logic with the
+		// higher layer network behaviour logic
+		SwarmBuilder::new(transport, behaviour, local_peer_id)
+			// connection background tasks are spawned onto the tokio runtime
+			.executor(Box::new(|fut| {
+				tokio::spawn(fut);
+			}))
+			.build()
```

Comment:
.max_negotiating_inbound_streams(10000) should prob be added to the swarm build.

i.e.:
```suggestion
		SwarmBuilder::new(transport, behaviour, local_peer_id)
			// connection background tasks are spawned onto the tokio runtime
			.executor(Box::new(|fut| {
				tokio::spawn(fut);
			}))
			.max_negotiating_inbound_streams(10000)
			.build()
```

Reply:
we should skip this, for now
also, there's a default value that we can use